### PR TITLE
UI, Radio Input

### PR DESCRIPTION
**Pull request by nhaagen**
This implements a radio-field with sub-forms dependant on options.

bildschirmfoto_2018-08-15_12-01-59

Rules are here: https://github.com/ILIAS-eLearning/ILIAS/pull/1133/files#diff-2d8571e3a451f51b51f9868081f6f7e8

### CODE REVIEW

File: src/UI/Component/Input/Field/Factory.php [hunk 1]
```diff
@@ -344,5 +344,37 @@ public function password($label, $byline = null);
 	 * @return \ILIAS\UI\Component\Input\Field\Select
 	 */
 	public function select($label, array $options, $byline = null);
+
+
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     A radio-input is used to depict a choice of options excluding each other.
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php]: Radio Input.

File: src/UI/Component/Input/Field/Factory.php [hunk 2]
```diff
@@ -344,5 +344,37 @@ public function password($label, $byline = null);
...
+	 *   composition: >
+	 *     The radio is considered as one field with a label and a number of
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php]: capitalize the name of the described component. Here Radio or Radio Input.

File: src/UI/Component/Input/Field/Factory.php [hunk 3]
```diff
@@ -344,5 +344,37 @@ public function password($label, $byline = null);
...
+	 *     options. Each option in turn bears a label in form of a positive statement.
+	 *   effect: >
+	 *     If used in a form, each option of a radio may open a dependant section (formerly known
+	 *     as sub form).
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php]: Again capitalize Radio, Dependant Section and even Sub Form.

File: src/UI/Component/Input/Field/Factory.php [hunk 4]
```diff
@@ -344,5 +344,37 @@ public function password($label, $byline = null);
...
+	 *   rivals:
+	 *     Checkbox field: Use a checkbox field for a binary yes/no choice.
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php]: Checkbox Field. Checkbox.

File: src/UI/Component/Input/Field/Factory.php [hunk 5]
```diff
@@ -344,5 +344,37 @@ public function password($label, $byline = null);
...
+	 *     Select: >
+	 *       Use selects to choose items from a longer list as the configuration of
+	 *       an aspect; when the choice has severe effects on, i.e. service behavior,
+	 *       or needs further cconfiguration, stick to radios.
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php]: cconfiguration.

File: src/UI/Component/Input/Field/Factory.php [hunk 6]
```diff
@@ -344,5 +344,37 @@ public function password($label, $byline = null);
...
+	 *
+	 * rules:
+	 *   wording:
+	 *     1: The options' labels MUST always state something positive.
+	 *
```
Comment (Amstutz) [inline src/UI/Component/Input/Field/Factory.php]: I have some more:

usage:

- A Radio Input SHOULD contain 3 to 5 options. If there are more, the Select Input might be the better option.
- They MAY also be used to select between two options where one is not automatically the inverse of the other (such as “Order by Date” and “Order by Name”).

wording:
1: Good, but give an example to make the point really clear here.
- Each option in a Radio Input MUST be labeled. 
 - This label SHOULD not consist of more than 5 words. 
- The label of the options SHOULD not simply repeat the label of the Radio. A meaningful labeling SHOULD be chosen instead.

Ordering:
- The presumably most relevant option SHOULD be the first option.

File: src/UI/Implementation/Component/Input/Field/Radio.php [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	/**
+	 * @var array <string,array> {$option_value => $fields}
+	 */
+	protected $dependant_fields = [];
+
+	/**
+	 * @inheritdoc
+	 */
+	protected function isClientSideValueOk($value) {
+		return true;
```
Comment (Amstutz) [inline src/UI/Implementation/Component/Input/Field/Radio.php]: Hmm, can we no do better here? Not sure however, give it a thougth if it's possible to restrict that somewhat more.

**Review by Amstutz — CHANGES_REQUESTED** [review]
Thanks a lot for this PR! I have some proposal for rules. Consider them, if you do not think them of any value, I can live with it. Some minor comments about capitalization.

Comment about the implementation: On Safari the checkboxes look really funny (large and ugly). However, you did not add any less, do you know why this is?

**Comment by nhaagen** [comment]
@Amstutz Thanks for looking through this and your suggestions. I'll merge the renderer and straighten out capitals ;)
The large input can also be found in checkboxes - this is a side effect of CSS class "form-control" with height:25px; I'll fix that, too.

**Comment by nhaagen** [comment]
This is much better now, in both description and looks...

**Comment by matthiaskunkel** [comment]
Jour Fixe, 27 AUG 2018 : We highly appreciate this PR and accept it for 5.4.

**Pull request merged**